### ISSUE TITLE
What's the semantics of `setXXXLE` methods for a little endian buffer?

### ISSUE THREAD

**Issue by zsxwing** [netty/netty#7465]
### Expected behavior

I'm implement the little endian methods (a.k.a `setXXXLE`). However, I feel confused about the semantics when the buffer is little endian.

When calling `setXXXLE` methods on a little endian buffer,  the current behavior is just the opposite of `setXXX` methods. However, the doc of `setXXXLE` methods says `Sets the specified 32-bit integer at the specified absolute index in this buffer with Little Endian byte order` and I read it as it should always write in Little Endian byte order.

Should the doc be improved to clarify the semantics for little endian buffer?

### Actual behavior

N/A

### Minimal yet complete reproducer code (or URL to code)

The following codes show the current behavior.

```Java
    @Test
    public void testSetIntLE() {
        ByteBuf b1 = ByteBufAllocator.DEFAULT.directBuffer();
        assert(b1.order() == ByteOrder.BIG_ENDIAN);
        ByteBuf b2 = b1.order(ByteOrder.LITTLE_ENDIAN);
        assert(b2.order() == ByteOrder.LITTLE_ENDIAN);
        b2.setIntLE(0, 1);
        assertEquals(0, b2.getByte(0));
        assertEquals(0, b2.getByte(1));
        assertEquals(0, b2.getByte(2));
        assertEquals(1, b2.getByte(3));
    }
```

### Netty version

4.1.x

### JVM version (e.g. `java -version`)

N/A

### OS version (e.g. `uname -a`)

N/A

**Comment by zsxwing** [comment]
@normanmaurer what do you think? Is it a doc issue or my understanding is wrong?

**Comment by rkapsi** [comment]
@zsxwing I think you're "double endianing" it and the JavaDoc is not helpful in that regard. First you create a ByteBuf that is little endian (i.e. all normal write ops should be little endian) but then you're doing an explicit little endian write which flips the bytes again.

I think the intention in respect to the API and what's deprecated is to not touch the `order(...)` and then use `get/set/write/readInt(...)` for big endian and `get/set/write/readIntLE(...)` for little endian.

```java
public class LittleEndianTest {

  @Test
  public void testNettyDirect() {
    nettyBuffer(ByteBufAllocator.DEFAULT.directBuffer(4, 4));
  }
  
  @Test
  public void testNettyHeap() {
    nettyBuffer(ByteBufAllocator.DEFAULT.heapBuffer(4, 4));
  }
  
  @Test
  public void testJdkDirect() {
    jdkBuffer(ByteBuffer.allocateDirect(4));
  }
  
  @Test
  public void testJdkHeap() {
    jdkBuffer(ByteBuffer.allocate(4));
  }
  
  private void nettyBuffer(ByteBuf buf) {
    assertTrue(buf.order() == ByteOrder.BIG_ENDIAN);
    
    ByteBuf dst = buf.order(ByteOrder.LITTLE_ENDIAN);
    assertTrue(dst.order() == ByteOrder.LITTLE_ENDIAN);
    
    dst.setInt(0, 1);
    
    assertEquals(1, dst.getByte(0));
    assertEquals(0, dst.getByte(1));
    assertEquals(0, dst.getByte(2));
    assertEquals(0, dst.getByte(3));
  }
  
  private void jdkBuffer(ByteBuffer buf) {
    assertTrue(buf.order() == ByteOrder.BIG_ENDIAN);
    
    ByteBuffer dst = buf.order(ByteOrder.LITTLE_ENDIAN);
    assertTrue(dst.order() == ByteOrder.LITTLE_ENDIAN);
    
    dst.putInt(0, 1);
    
    assertEquals(1, dst.get(0));
    assertEquals(0, dst.get(1));
    assertEquals(0, dst.get(2));
    assertEquals(0, dst.get(3));
  }
}
```

**Comment by zsxwing** [comment]
@rkapsi Yeah, I already took a look at the implementation and understand what's happening internally.

However, as a Netty user, how to implement `readInt` and `readIntLE` for a LE buffer? I think the behavior should be clear in the doc. Right now it's really confusing. Should I make my LE buffer have the same behavior like a buffer creating using `bigEndianBuffer.order(ByteOrder.LITTLE_ENDIAN);` (a.k.a, the deprecated SwappedByteBuf)? (this is what I did in https://github.com/apache/arrow/pull/1376) If so, the method name and doc is really misleading.

**Comment by rkapsi** [comment]
@zsxwing understood. My take would be to not use `order(...)` because it's deprecated and if I did then I'd use the "endian neutral" `setXxx` methods (the equivalents of JDK's ByteBuffer `putXxx`).

If I didn't have a legacy code then I'd ignore `order(...)`. Maybe add JavaDoc regards not using `order(...)` in conjunction with `setXxxLE`?!

**Comment by zsxwing** [comment]
@rkapsi I'm not using `order`. Apache Arrow implements its own ByteBuf and it's in LE byte order. Now I'm upgrading Netty for `Apache Arrow` and have to implement these new setXxxLE methods. Therefore, I'm wondering what's the correct semantics to implement.
